crates/evm/core/src/backend/code_cache.rs

```diff
@@ -0,0 +1,150 @@
+use alloy_chains::Chain;
+use alloy_provider::{Network, Provider};
+use alloy_transport::{Transport, TransportResult};
+use quick_cache::sync::Cache;
+use revm::primitives::{Address, Bytes};
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+struct CodeDetected {
+    /// The block number at which the code was detected.
+    block_number: u64,
+    /// The code at that block.
+    code: Bytes,
+}
+
+#[derive(Debug, Default, Clone)]
+struct CodeHistory {
+    /// The block number at which the code was first detected. with the code at that block.
+    code_first_detected_at: Option<CodeDetected>,
+    /// The block number at which the eoa was last detected.
+    eoa_last_detected_at: Option<u64>,
+}
+
+#[derive(Debug)]
+pub struct CodeCache(Cache<(Address, Chain), CodeHistory>);
+
+impl Default for CodeCache {
+    fn default() -> Self {
+        Self(Cache::new(10_000))
+    }
+}
+
+impl CodeCache {
+    /// Get the code of an account at a specific block, using the cache if possible.
+    /// If the code is not in the cache, it will be fetched from the provider and cached.
+    pub async fn get_code<N: Network, T: Transport + Clone, P: Provider<T, N>>(
+        &self,
+        provider: &P,
+        address: Address,
+        chain: Chain,
+        block_number: u64,
+    ) -> TransportResult<Bytes> {
+        if let Some(code) = self.check_cache(address, chain, block_number) {
+            return Ok(code);
+        }
+
+        let code = provider.get_code_at(address, block_number.into()).await?;
+
+        self.cache_code(address, chain, block_number, code.clone());
+
+        Ok(code)
+    }
+
+    /// Check the cache for the code of an account at a specific block.
+    /// Returns the code if it is in the cache, otherwise None.
+    ///
+    /// If the account had code at the time of the block or earlier, it had code at the time of the
```

Comment:
This comment is pretty hard to comprehend, and the comments on the CodeHistory struct also don't describe why they are Option and what the states might mean. Can you expand on them?

Reply:
Updated type defs and comments, lmk if still confusing.